executor/common_linux.h

```diff
@@ -4103,3 +4103,281 @@ static void setup_usb()
 // The macro is used in generated C code.
 #define CAST(f) ({void* p = (void*)f; p; })
 #endif
+
+#if SYZ_EXECUTOR || __NR_syz_mount_fuse
+#include <fcntl.h>
+#include <linux/fuse.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+/* Perform correct FUSE initialization:
+ * 1) Create target directory if missing
+ * 2) Mount FUSE fs on top of target directory
+ * 3) Return fd referring to the target directory
+ */
+static volatile long syz_mount_fuse(volatile long a0, /* /dev/fuse fd. */
+				    volatile long a1, /* Mount target. */
+				    volatile long a2) /* Mount flags. */
+{
+	const char* file_system_type = "fuse";
+	const char* mount_source = "fuse"; /* Not used by FUSE. */
+	char* mount_target;
+	char opts[128];
+	unsigned long mount_flags;
+	struct stat stbuf;
+	int fd_fuse;
+	int ret;
+
+	fd_fuse = (int)a0;
+	mount_target = (char*)a1;
+	mount_flags = (unsigned long)a2;
+
+	ret = mkdir(mount_target, S_IRWXU);
+	if (ret == -1) {
+		debug("syz_mount_fuse() > mkdir() failed: %d\n", errno);
+		return -1;
+	}
+
+	/* Retrieve mount_target file type. We could probably hardcode S_IFDIR,
+         * since we always create a directory, but this will keep things working
+         * if we start fuzzing the file type.
+         * */
+	ret = stat(mount_target, &stbuf);
+	if (ret == -1) {
+		debug("syz_mount_fuse() > stat() failed: %d\n", errno);
+		return -1;
+	}
+
+	/* Currently we do not support any optional settings, because they can
+         * be tested/fuzzed through the mount$fuse() syscall.
+         */
+	snprintf(opts, sizeof(opts), "fd=%i,rootmode=%o,user_id=%u,group_id=%u",
+		 fd_fuse, stbuf.st_mode & S_IFMT, getuid(), getgid());
+	ret = mount(mount_source, mount_target, file_system_type, mount_flags, opts);
+	if (ret) {
+		debug("syz_mount_fuse() > mount() failed: %d\n", errno);
+		return -1;
+	}
+
+	return open(mount_target, O_RDONLY | O_DIRECTORY);
+}
+#endif
+
+#if SYZ_EXECUTOR || __NR_syz_open_fuse_file
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+static volatile long syz_open_fuse_file(volatile long a0, /* Dir fd. */
+					volatile long a1, /* Filename. */
+					volatile long a2) /* Open flags. */
+{
+	char* filename;
+	int open_flags;
+	int dir_fd;
+	int fd;
+
+	dir_fd = (int)a0;
+	filename = (char*)a1;
+	open_flags = ((int)a2) | O_CREAT;
+	fd = openat(dir_fd, filename, open_flags);
```

Comment:
Pseudo-syscalls is a last resort cases that we can't handle in descriptions and even can't handle any approximation of what we need.
This pseudo-syscall does not seem to do anything that can't be done in descriptions, use openat syscall specialization in descriptions instead.

Reply:
Is it possible to force the O_CREAT flag while fuzzing the remaining flags?

I had initially tested using a specialized syscall and this didn't seem to be the case (and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage).

Reply:
See `syzkaller$ grep "define .*|" sys/linux/*.txt`
But why do we want O_CREAT always? I think we want to open existing files equally, no?

> and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage

Testing fuzzer changes is tricky. What was the methodology for testing?

Reply:
> But why do we want O_CREAT always? I think we want to open existing files equally, no?

O_CREAT just says create if missing, it still allows opening existing files. To test FUSE, we are mounting a brand new filesystem every time and therefore almost always openat() is targeting nonexistent files (this is why I would like to always set O_CREAT for this specific use case).

> Testing fuzzer changes is tricky. What was the methodology for testing?

I have come to realize that, haha. The best I have come up with is letting it run for 1-3 days on 12 VMs with 2 CPUs each (~ 300-900 hours of fuzzing). Is there a better way?

Reply:
Right, I confused it with O_EXCL.
Every test starts with an empty dir every time, so this looks like non-fuse specific and affects all open calls.

> I have come to realize that, haha. The best I have come up with is letting it run for 1-3 days on 12 VMs with 2 CPUs each (~ 300-900 hours of fuzzing). Is there a better way?

No :) But at least it's not a 1-hour run.

Reply:
> I had initially tested using a specialized syscall and this didn't seem to be the case (and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage).

After testing again, files are being opened correctly using the normal openat() syscall (as one would expect). I think I was defining the openat$fuse_file(...) variant with 3 parameters (no mode) and that was somehow messing things up.